clang/test/CodeGen/logb_scalbn.c

```diff
@@ -0,0 +1,237 @@
+// RUN: %clang -cc1 -triple amdgcn-amd-amdhsa -emit-llvm -o - %s | FileCheck %s
+
+// CHECK-LABEL: define dso_local void @test_logb(
+// CHECK-SAME: ) #[[ATTR0:[0-9]+]] {
+// CHECK-NEXT:  [[ENTRY:.*:]]
+// CHECK-NEXT:    [[D1:%.*]] = alloca float, align 4, addrspace(5)
+// CHECK-NEXT:    [[D1_ASCAST:%.*]] = addrspacecast ptr addrspace(5) [[D1]] to ptr
+// CHECK-NEXT:    [[TMP0:%.*]] = call { double, i32 } @llvm.frexp.f64.i32(double 0x40301999A0000000)
+// CHECK-NEXT:    [[TMP1:%.*]] = extractvalue { double, i32 } [[TMP0]], 1
+// CHECK-NEXT:    [[TMP2:%.*]] = add nsw i32 [[TMP1]], -1
+// CHECK-NEXT:    [[TMP3:%.*]] = sitofp i32 [[TMP2]] to double
+// CHECK-NEXT:    [[TMP4:%.*]] = call double @llvm.fabs.f64(double 0x40301999A0000000)
+// CHECK-NEXT:    [[TMP5:%.*]] = fcmp one double [[TMP4]], 0x7FF0000000000000
+// CHECK-NEXT:    [[TMP6:%.*]] = select i1 [[TMP5]], double [[TMP3]], double [[TMP4]]
+// CHECK-NEXT:    [[TMP7:%.*]] = select i1 false, double 0xFFF0000000000000, double [[TMP6]]
+// CHECK-NEXT:    [[CONV:%.*]] = fptrunc double [[TMP7]] to float
+// CHECK-NEXT:    store float [[CONV]], ptr [[D1_ASCAST]], align 4
+// CHECK-NEXT:    ret void
+//
+void test_logb() {
+  float D1 = __builtin_logb(16.1f);
+}
+// CHECK-LABEL: define dso_local void @test_logb_var(
+// CHECK-SAME: float noundef [[A:%.*]]) #[[ATTR0]] {
+// CHECK-NEXT:  [[ENTRY:.*:]]
+// CHECK-NEXT:    [[A_ADDR:%.*]] = alloca float, align 4, addrspace(5)
+// CHECK-NEXT:    [[D1:%.*]] = alloca float, align 4, addrspace(5)
+// CHECK-NEXT:    [[A_ADDR_ASCAST:%.*]] = addrspacecast ptr addrspace(5) [[A_ADDR]] to ptr
+// CHECK-NEXT:    [[D1_ASCAST:%.*]] = addrspacecast ptr addrspace(5) [[D1]] to ptr
+// CHECK-NEXT:    store float [[A]], ptr [[A_ADDR_ASCAST]], align 4
+// CHECK-NEXT:    [[TMP0:%.*]] = load float, ptr [[A_ADDR_ASCAST]], align 4
+// CHECK-NEXT:    [[CONV:%.*]] = fpext float [[TMP0]] to double
+// CHECK-NEXT:    [[TMP1:%.*]] = call { double, i32 } @llvm.frexp.f64.i32(double [[CONV]])
+// CHECK-NEXT:    [[TMP2:%.*]] = extractvalue { double, i32 } [[TMP1]], 1
+// CHECK-NEXT:    [[TMP3:%.*]] = add nsw i32 [[TMP2]], -1
+// CHECK-NEXT:    [[TMP4:%.*]] = sitofp i32 [[TMP3]] to double
+// CHECK-NEXT:    [[TMP5:%.*]] = load float, ptr [[A_ADDR_ASCAST]], align 4
+// CHECK-NEXT:    [[CONV1:%.*]] = fpext float [[TMP5]] to double
+// CHECK-NEXT:    [[TMP6:%.*]] = call double @llvm.fabs.f64(double [[CONV1]])
+// CHECK-NEXT:    [[TMP7:%.*]] = fcmp one double [[TMP6]], 0x7FF0000000000000
+// CHECK-NEXT:    [[TMP8:%.*]] = select i1 [[TMP7]], double [[TMP4]], double [[TMP6]]
+// CHECK-NEXT:    [[TMP9:%.*]] = fcmp oeq double [[CONV]], 0.000000e+00
+// CHECK-NEXT:    [[TMP10:%.*]] = select i1 [[TMP9]], double 0xFFF0000000000000, double [[TMP8]]
+// CHECK-NEXT:    [[CONV2:%.*]] = fptrunc double [[TMP10]] to float
+// CHECK-NEXT:    store float [[CONV2]], ptr [[D1_ASCAST]], align 4
+// CHECK-NEXT:    ret void
+//
+void test_logb_var(float a) {
+  float D1 = __builtin_logb(a);
+}
```

Comment:
Testing the implicit cast isn't useful, the flat versions should test the f suffixed versions